Simplify and centralize the processing of internal states for TreeSelectControl

eason9487 opened the pull request:
### Changes proposed in this Pull Request:

To simplify and centralize the processing of updating expanded, checked, and partially checked states, this PR:
- Replace `cloneDeep` with array `reduce` when generating filtered options.
- Calculate the expanded, checked, and partially checked states of options in the main component.
- Move the processing of updating the expanded state to the main component.
- Simplify the handling of updating the state of selected values.

### Detailed test instructions:

1. Go to the Edit free listings or the local storybook.
2. Test if the states and functionality of checking/unchecking options work correctly.
3. Test if the states and functionality of expanding/collapsing options work correctly.

### Changelog entry

> Tweak - Simplify and centralize the processing of internal states for the TreeSelectControl component.

eason9487 committed:
Replace cloneDeep with reduce when generating filtered options in TreeSelectControl

eason9487 committed:
Calculate the expanded state of options in the main component of TreeSelectControl

eason9487 committed:
Move the handling of updating the expanded state to the main component of TreeSelectControl.

eason9487 committed:
Calculate the checked and partially checked states of options in the main component of TreeSelectControl

eason9487 committed:
Simplify the handling of updating the selected values in TreeSelectControl

puntope commented on js/src/components/tree-select-control/index.js:
```diff
@@ -92,14 +110,16 @@ const TreeSelectControl = ( {
 	const onDropdownVisibilityChangeRef = useRef();
 	onDropdownVisibilityChangeRef.current = onDropdownVisibilityChange;
 
 	// We will save in a REF previous search filter queries to avoid re-query the tree and save performance
 	const filteredOptionsCache = useRef( {} );
+	const cacheStatesRef = useRef( {} );
+	cacheStatesRef.current.expandedValues = nodesExpanded;
+	cacheStatesRef.current.selectedValues = value;
```
💅  WDTY of saving all the caches in

`cacheRef.current.filteredOptions`
`cacheRef.current.expandedValues `
`cacheRef.current.selectedValues `

eason9487 replied:
Changed in https://github.com/woocommerce/google-listings-and-ads/pull/1522/commits/e2a6a827289312f796e202e68b628ed34f0485d4.

puntope commented on js/src/components/tree-select-control/index.js:
```diff
@@ -140,23 +159,29 @@ const TreeSelectControl = ( {
+			return cachedFilteredOptions;
+		}
+
+		const isFiltered = Boolean( filter );
```
❓  Why is this necessary? 

Just using `filter` should work I guess

eason9487 replied:
In this context, `isFiltered` is more semantic than `filter`, and the explicit boolean type allows the below return value to always be a boolean type.

https://github.com/woocommerce/google-listings-and-ads/blob/afa8de0035bd66217caa96b5212fb51ff1ea2be9/js/src/components/tree-select-control/index.js#L258-L266

puntope replied:
When I saw it I thought for a moment that was kind of a flag indicating that a node was filtered or not. 

💅  Maybe consider to change it to `hasFilters`

eason9487 replied:
> When I saw it I thought for a moment that was kind of a flag indicating that a node was filtered or not.

The naming was taken from the former prop used in the Options component.

https://github.com/woocommerce/google-listings-and-ads/blob/65d232a5aab91dad7910ec352a6389ff4c6f140c/js/src/components/tree-select-control/options.js#L26

> Maybe consider to change it to `hasFilters`

How about `isSearching`?

puntope replied:
isSearching looks pretty accurate for me yes! 🤝

eason9487 replied:
Changed in https://github.com/woocommerce/google-listings-and-ads/pull/1522/commits/6b13282a3297613e3d68cd3589315b9c6721e339.

puntope commented on js/src/components/tree-select-control/index.js:
```diff
@@ -169,30 +194,104 @@ const TreeSelectControl = ( {
-			setFilteredOptions( cachedFilteredOptions );
-			return;
-		}
+				if ( option.children.length ) {
+					Object.defineProperties( option, descriptors );
+					acc.push( option );
+				}
+			} else {
+				if ( isFiltered ) {
+					const match = option.label.toLowerCase().indexOf( filter );
+					if ( match === -1 ) {
+						return acc;
+					}
+					option.label = highlightOptionLabel( option.label, match );
+				}
+
+				Object.defineProperties( option, descriptors );
+				acc.push( option );
+			}
```
💅 

Just an idea, I guess the reducer could be simplified with: 

```
const descriptors = {
            hasChildren: {
				get() {
					return this.children?.length > 0;
				},
			},
			allLeaves: {
```
```
const reduceOptions = ( acc, { children = [], ...option } ) => {
			if ( children.length ) {
				option.children = children.reduce( reduceOptions, [] );
			} else if ( isFiltered ) {
				const match = option.label.toLowerCase().indexOf( filter );
				if ( match === -1 ) {
					return acc;
				}
				option.label = highlightOptionLabel( option.label, match );
			}

			if ( option.children?.length || ! children.length ) {
				Object.defineProperties( option, descriptors );
				acc.push( option );
			}
			return acc;
		};
```

puntope replied:
💅 WDYT of adding a small comment in the reducer?

Something like "This reducer gets the options matching the filter and also adds Leaf properties to them"

eason9487 replied:
Thanks for the nice suggestion! Based on your example, I adjusted it to a similar implementation in https://github.com/woocommerce/google-listings-and-ads/pull/1522/commits/da745d8b36ba4e9e020ae00604f85d8fcf238630.

---

> WDYT of adding a small comment in the reducer?

The processing of this reducer is also included in the code comments above, so I guess it's already sufficient?

https://github.com/woocommerce/google-listings-and-ads/blob/afa8de0035bd66217caa96b5212fb51ff1ea2be9/js/src/components/tree-select-control/index.js#L163-L169

puntope commented on js/src/components/tree-select-control/index.js:
```diff
@@ -46,13 +46,31 @@ import { ARROW_DOWN, ARROW_UP, ENTER, ESCAPE, ROOT_VALUE } from './constants';
+ * @typedef {Object} CommonOption
  * @property {string} value The value for the option
+ * @property {string} [key] Optional unique key for the Option. It will fallback to the value property if not defined
+ */
+
+/**
+ * @typedef {Object} BaseOption
  * @property {string} label The label for the option
  * @property {Option[]} [children] The children Option objects
- * @property {string} [key] Optional unique key for the Option. It will fallback to the value property if not defined
- **/
+ *
+ * @typedef {CommonOption & BaseOption} Option
+ */
```
💅  I will use just `Option` instead of CommonOption and BaseOption. (Or maybe I missed the intent of this separation). 

@typedef {Object} Option
@property {string} label The label for the option
@property {string} value The value for the option
@property {string} [key] Optional unique key for the Option. It will fallback to the value property if not defined
@property {Option[]} [children] The children Option objects

💅  WDYT of renaming `InnerOption` as well as `BaseInnerOption` to `Leaf` and `BaseLeaf`

In my mind I see it this way:
Option is the "pure" item we send to the tree, the data itself, with label, value, and optional key and children.
Leaf is and enhanced Option with Tree props (BaseLeaf).

eason9487 replied:
> I will use just `Option` instead of CommonOption and BaseOption. (Or maybe I missed the intent of this separation).

Option and InnerOption both have `label` and `children` properties, but the type and description of the properties are different. JSDoc does not override the same attributes when using `@typedef {Option & BaseInnerOption} InnerOption`, but combines them.

For example, using this JSDoc: 

```js
/**
 * @typedef {Object} Option
 * @property {string} label The label for the option.
 * @property {string} value The value for the option.
 * @property {string} [key] Optional unique key for the Option. It will fallback to the value property if not defined.
 * @property {Option[]} [children] The children Option objects.
 */

/**
 * @typedef {Object} BaseInnerOption
 * @property {string|JSX.Element} label The label string or label with highlighted react element for the option.
 * @property {InnerOption[]|undefined} children The children options. The options are filtered if in searching.
 * (omitted)
 * @typedef {Option & BaseInnerOption} InnerOption
 */
```

There will be a combined type and description:

![2022-05-25 12 15 46](https://user-images.githubusercontent.com/17420811/170179900-3022b31f-2a00-450e-ba3f-d8e3f9e2c9a9.png)

Furthermore, the nested reference produces inconsistent results.

![2022-05-25 12 19 40](https://user-images.githubusercontent.com/17420811/170179936-1238fa78-4a9b-48a9-b694-a3414b201dc1.png)
![2022-05-25 12 21 10](https://user-images.githubusercontent.com/17420811/170179947-bb4b5b8e-e9ec-4699-8949-87e95c328f18.png)

---

> WDYT of renaming `InnerOption` as well as `BaseInnerOption` to `Leaf` and `BaseLeaf`

In a tree data structure, the terminology of *leaf* node refers to the node with no children. But an `InnerOption` could be a parent node, both parent and child node, or a leaf node. Therefore, I don't think `Leaf` is a proper name.

puntope replied:
> In a tree data structure, the terminology of leaf node refers to the node with no children. But an InnerOption could be a parent node, both parent and child node, or a leaf node. Therefore, I don't think Leaf is a proper name.

👌  Right. Maybe `node` then. 

> There will be a combined type and description:

Interesting. In my IDE it does the intersection properly.

<img width="727" alt="Screenshot 2022-05-25 at 12 31 34" src="https://user-images.githubusercontent.com/5908855/170218293-6442b841-7daa-4d40-a3f4-b325ba376cf2.png">

My worries are that Inner, Base, Common prefixes could be kinda confusing but this is a non blocking issue.

eason9487 replied:
> Maybe `node` then.

`Node` is a more general and abstract naming. When viewing the code like `@param {Node} option`, I don't think it's clearer, and it also seems more irrelevant to the parameter name.

> My worries are that Inner, Base, Common prefixes could be kinda confusing

Sorry, I'm not quite sure how they cause confusion in the scope of these type definitions since the Base and Common prefixes are only used for combining the Option and InnerOption. But I'd like to know if you have other naming ideas to replace them to reduce your confusion.

puntope replied:
All the names I have in mind rename Option to Node for defining the items of the tree and differentiate them from the Options provided by the user.

Option = { OptionValue & OptionData } 
Node = { OptionValue & BaseNode } 

`<Options options="filteredOptions"> to be renamed to <Nodes nodes="filteredNodes">`

But that requires more refactor of the Options component that was named wrongly from scratch by me. (notice in the Options.js component we refer to Nodes in ArrowLeft/ArrowRight docs) 

But this is pretty much a subjective and minor issue (unsure even if it worth it all the refactor in renaming).

eason9487 replied:
I think the overall naming adjustments could be a separate PR. Together with, for example, the `value` prop could be internally renamed to `values` or `selectedValues`, and the `nodesExpanded` to `expandedValues`, etc.

puntope commented on js/src/components/tree-select-control/index.js:
```diff
@@ -169,30 +194,104 @@ const TreeSelectControl = ( {
-				return !! option.children.length;
-			}
+		const descriptors = {
+			allLeaves: {
```
💅  I will keep it as `leaves` instead of `allLeaves`. When I saw `all` highlighted in the name of the prop, I was expecting to have just a generic `leaves` somewhere in the code. 

Also the description

`@property {InnerOption[]} allLeaves All (filtered) leaf options under this option.`

I'd suggest to replace it with something like: 

`@property {Leaf[]} leaves All the filtered option children flattened.`

PD. I guess InnerOption is not accurate since we have for example the option.label as well??? 🤔

eason9487 replied:
Renamed the `allLeaves` and adjusted some JSDocs in https://github.com/woocommerce/google-listings-and-ads/pull/1522/commits/afa8de0035bd66217caa96b5212fb51ff1ea2be9.

> PD. I guess InnerOption is not accurate since we have for example the option.label as well???

Sorry, I don't quite understand this. The `InnerOption` does have the `label` property. And what does *PD.* refer to? Could you elaborate more?

puntope replied:
> Sorry, I don't quite understand this. The InnerOption does have the label property. And what does PD. refer to? Could you elaborate more?

Apologies. I missed the label property inside `InnerOption`. All good 👌 (with PD I meant PS, had a typo)

puntope reviewed (commented):
Howdy @eason9487 

I did a first check for this PR and it works amazing, also it simplifies a lot the relationship between the Options component and the tree ❤️ 

In left some comments tho which I think deserve discussion.

Thanks!

eason9487 committed:
Use the same ref to keep the caches for TreeSelectControl.

Address https://github.com/woocommerce/google-listings-and-ads/pull/1522#discussion_r880253033

eason9487 committed:
Further simplify the `reduceOptions` function in TreeSelectControl.

Address https://github.com/woocommerce/google-listings-and-ads/pull/1522#discussion_r880279233

eason9487 committed:
Rename the `allLeaves` to `leaves` and adjust some JSDocs in TreeSelectControl.

Address https://github.com/woocommerce/google-listings-and-ads/pull/1522#discussion_r880302895

puntope commented on js/src/components/tree-select-control/index.js:
```diff
@@ -169,30 +194,104 @@ const TreeSelectControl = ( {
-				return !! option.children.length;
-			}
+		const descriptors = {
+			hasChildren: {
+				get() {
+					return this.children?.length > 0;
+				},
+			},
```
💅  Missing docs

eason9487 replied:
Added in https://github.com/woocommerce/google-listings-and-ads/pull/1522/commits/6dfadfc6f50e9938f0f5434a1b4fdad1a9bcf93c.

puntope reviewed (approved):
✅  All the discussion were addressed! This is good to go. Thanks @eason9487 

I left some opinionated comments 💅  that could improve even more the code IMO. Those are optional or could be done in future PR. So this is approved.

eason9487 committed:
Rename `isFiltered` to `isSearching` in TreeSelectControl.

Address https://github.com/woocommerce/google-listings-and-ads/pull/1522#discussion_r881353563

eason9487 committed:
Add JSDoc for the `hasChildren` function in TreeSelectControl.

Address https://github.com/woocommerce/google-listings-and-ads/pull/1522#discussion_r881359742

eason9487 committed:
Rename the `current` of destructuring assignment to `cache` in TreeSelectControl.

Address https://github.com/woocommerce/google-listings-and-ads/pull/1522#commitcomment-74459851
